### PR TITLE
Dummy PR for review comments

### PR DESCRIPTION
**Pull request by sbihel**

### CODE REVIEW

File: siwe.go [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+)
+
+type MalformedMessage struct{ string }
+type ParsingFailed struct{ string }
+type ExpiredMessage struct{ string }
+type InvalidMessage struct{ string }
+type InvalidSignature struct{ string }
+
+func (m *ParsingFailed) Error() string {
+	return "Expired Message"
```
Comment (sbihel) [inline siwe.go:24]: Wrong message. Also what's the difference between `ParsingFailed` and `InvalidMessage`?

File: siwe.go [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+type Message struct {
+	domain  string
+	address common.Address
+	uri     string
+	version string
+
+	statement *string
+	nonce     *string
+	chainID   string
```
Comment (sbihel) [inline siwe.go:47]: Should be an int

File: siwe.go [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return parsed, nil
+}
+
+func signHash(data []byte) common.Hash {
```
Comment (sbihel) [inline siwe.go:343]: Should be a method of `Message`, and probably call it `eip191Hash` like in `siwe-rs`

File: siwe.go [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+	address := crypto.PubkeyToAddress(*pkey)
+
+	if address != m.address {
+		return nil, &InvalidSignature{"Signer address must match message address"}
+	}
+
+	return pkey, nil
+}
+
+func (m *Message) Verify(signature string) (*ecdsa.PublicKey, error) {
```
Comment (sbihel) [inline siwe.go:402]: Following yesterday's discussion, should accept expected nonce, timestamp, etc. 

And I don't think siwe-rs made the right choice with exposing `verify_eip191`.

File: siwe.go [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+	hash := signHash([]byte(data))
+
+	sigBytes, err := hexutil.Decode(signature)
+	if err != nil {
+		return nil, &InvalidSignature{"Failed to decode signature"}
+	}
+
+	// Ref:https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L442
+	if sigBytes[64] != 27 && sigBytes[64] != 28 {
+		return nil, &InvalidSignature{"Invalid signature bytes"}
```
Comment (sbihel) [inline siwe.go:384]: I would rephrase the message to `Invalid signature recovery byte`

File: siwe.go [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+}
+
+type Message struct {
+	domain  string
+	address common.Address
+	uri     string
+	version string
+
+	statement *string
+	nonce     *string
```
Comment (sbihel) [inline siwe.go:46]: I think it might be preferable to not have it optional, if a user wants to build a message they should create a nonce and be forced to store it somewhere for verification later.

File: siwe.go [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+	_, err := m.ValidNow()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return m.VerifyEIP191(signature)
+}
+
+func (m *Message) PrepareMessage() string {
```
Comment (sbihel) [inline siwe.go:412]: I would make it the built in `String` method

File: siwe_test.go [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+		var result map[string]interface{}
+		err := json.Unmarshal([]byte(data), &result)
+		assert.Nil(t, err)
+
+		switch test {
+		case "parsing-negative":
+			parsingNegative(t, result)
+		case "parsing-positive":
+			parsingPositive(t, result)
```
Comment (sbihel) [inline siwe_test.go:338]: You can use those tests to test the serialisation as well, by checking the round trip eip4361 -> object -> eip4361

File: siwe.go [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+	}
+	return nil
+}
+
+func (m *Message) GetResources() []string {
+	return m.resources
+}
+
+func GenerateNonce() string {
+	return uniuri.NewLen(16)
```
Comment (sbihel) [inline siwe.go:254]: That package generates random strings for uniqueness, not cryptographically sound nonces

File: siwe.go [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+			}
+			value = val.(string)
+		default:
+			return nil, &InvalidMessage{"`expirationTime` must be either an ISO8601 formatted string or time.Time"}
+		}
+		expirationTime = &value
+	}
+
+	var notBefore *string
+	if val, ok := options["notBefore"]; ok {
```
Comment (sbihel) [inline siwe.go:117]: Looks like this code could be factored out to avoid duplication.

File: siwe.go [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+}
+
+func (m *InvalidSignature) Error() string {
+	return fmt.Sprintf("Invalid Signature: %s", m.string)
+}
+
+type Message struct {
+	domain  string
+	address common.Address
+	uri     string
```
Comment (sbihel) [inline siwe.go:42]: Looks like there are packages that offer a URI type

File: siwe.go (hunk 1 above)
Comment (theosirian) [reply siwe.go:24]: I was using it to describe the regex parsing error, but it doesn't make much sense, everywhere else in field validation I was using InvalidMessage, so I'm gonna keep just that one.

File: siwe.go (hunk 4 above)
Comment (theosirian) [reply siwe.go:402]: Timestamp makes sense to accept in this case, but why the nonce? It should already be checked against the internal nonce field in the message, since that is what will be used to generate the hash to compare the signature against.

File: siwe.go (hunk 7 above)
Comment (theosirian) [reply siwe.go:412]: Should we keep PrepareMessage as well or prefer each language's toString equivalent exclusively?

File: siwe.go (hunk 9 above)
Comment (theosirian) [reply siwe.go:254]: I'm looking for another alternative, but it seems to be basically the same thing we do in siwe-rs .
```
    thread_rng()
        .sample_iter(&Alphanumeric)
        .take(11)
        .map(char::from)
        .collect()
```

File: siwe.go (hunk 11 above)
Comment (theosirian) [reply siwe.go:42]: I'm going to try to change it to an URI, but one problem might be that after parsing, if the String()method of the URI doesn't produce the exactly same string we will fail verification.

What I've done for timestamps is try to parse them into a `time.Time` object and keep the string if successful.

File: siwe.go (hunk 4 above)
Comment (sbihel) [reply siwe.go:402]: But the Message is what was signed, not what the server wanted to user to sign. I don't think the server should try to rebuild the message and only receive the signature, because e.g. it cannot now the timestamp of when the message was signed, or the resources that ended up in the message. 

So the flow should be, have the server use both the message and signature sent by the frontend (or whatever else) and protect against basic attacks.

File: siwe.go (hunk 7 above)
Comment (sbihel) [reply siwe.go:412]: I think `String` can replace the current `PrepareMessage` because it only serialises the message. It's a bit tricky when you want to provide an interface to generate messages to be signed but we don't have to care about that for now.

File: siwe.go (hunk 9 above)
Comment (sbihel) [reply siwe.go:254]: My bad, it looks sound, I got confused with their focus on UUIDs

File: siwe.go (hunk 11 above)
Comment (sbihel) [reply siwe.go:42]: For URI it shouldn't be a problem. But yeah for timestamps it's a bit of a mess :pensive: parsing just for verification is fine, we can get more fancy if need be but that's good for now.

File: siwe.go (hunk 4 above)
Comment (theosirian) [reply siwe.go:402]: I understand what you are saying, but this doesn't prevent that use case. The message as sent by the user will be parsed into the `Message` object and then when `m.Verify` is called with the signature, the nonce and timestamp are already set to either the server's version or the user's, depending on the use case.

File: siwe.go (hunk 6 above)
Comment (theosirian) [reply siwe.go:46]: That was my mistake on typing, it's a pointer, but it's not actually optional, in the case the user doesn't provide one, one will be automatically generated and added to the message. I see now that this might prevent an user from verifying a message that doesn't have a `nonce`, since it's impossible to construct one without with this library.
We can either:
- Keep adding a nonce if not present in the initialization options; or
- Throw an error if not present.
Besides that, I've fixed the type in the review branch.

Comment (sbihel) [reply siwe.go:46]: I think probably throw an error. I suppose ideally we wouldn't have a builder method and just let the user populate the struct, but it could be an issue due to loose type.

File: siwe.go (hunk 4 above)
Comment (theosirian) [reply siwe.go:402]: The parsed message would already have the correct `nonce`, no? Since we'd be parsing the front-end provided text message.

Comment (sbihel) [reply siwe.go:402]: Yeah but the correct nonce is the one stored in the backend

**Comment by sbihel** [comment]
Closed in #11
